upcoming:[DI-24112] - Add grouping by tag for alerts

ankita-akamai opened the pull request:
## Description 📝

Highlight the Pull Request's context and intentions.

## Changes  🔄

List any change(s) relevant to the reviewer.

- ...
- ...

## Target release date 🗓️

Please specify a release date (and environment, if applicable) to guarantee timely review of this PR. If exact date is not known, please approximate and update it as needed.

## Preview 📷

**Include a screenshot or screen recording of the change.**

:lock: Use the [Mask Sensitive Data](https://cloud.linode.com/profile/settings) setting for security.

:bulb: Use `<video src="" />` tag when including recordings in table.

| Before  | After   |
| ------- | ------- |
| No grouping | ![image](https://github.com/user-attachments/assets/480dd354-8a9f-440e-82b5-56f08cfd4df8) |
| No grouping | ![image](https://github.com/user-attachments/assets/dd692373-012c-4d51-95ff-5ef2d89c52b9) |
| No grouping | ![image](https://github.com/user-attachments/assets/6fa66981-9add-4cf1-9f57-14bf03c4a13a) |

Mobile:
|||
| ------- | ------- |
| No grouping | ![image](https://github.com/user-attachments/assets/d4b0e80b-cea0-487f-aca7-d6355836beb5) |

## How to test 🧪

### Prerequisites

(How to setup test environment)

- ...
- ...

### Reproduction steps

(How to reproduce the issue, if applicable)

- [ ] ...
- [ ] ...

### Verification steps

(How to verify changes)

- [ ] ...
- [ ] ...

<details>
<summary> Author Checklists </summary>

## As an Author, to speed up the review process, I considered 🤔

*Check all that apply*

- [ ] Use React components instead of HTML Tags
- [ ] Proper naming conventions like cameCase for variables & Function & snake_case for constants
- [ ] Use appropriate types & avoid using "any"
- [ ] No type casting & non-null assertions
- [ ] Adding a [changeset](https://github.com/linode/manager/blob/develop/docs/CONTRIBUTING.md#writing-a-changeset)
- [ ] Providing/Improving test coverage
- [ ] Use sx props to pass styles instead of style prop
- [ ] Add JSDoc comments for interface properties & functions
- [ ] Use strict equality (===) instead of double equal (==)
- [ ] Use of named arguments (interfaces) if function argument list exceeds size 2
- [ ] Destructure the props
- [ ] Keep component size small & move big computing functions to separate utility
- [ ] 📱 Providing mobile support

<br/>

- [ ] I have read and considered all applicable items listed above.

## As an Author, before moving this PR from Draft to Open, I confirmed ✅

- [ ] All unit tests are passing
- [ ] TypeScript compilation succeeded without errors
- [ ] Code passes all linting rules

</details>

---

## Commit message and pull request title format standards

> **Note**: Remove this section before opening the pull request
**Make sure your PR title and commit message on squash and merge are as shown below**

`<commit type>: [JIRA-ticket-number] - <description>`

**Commit Types:**

- `feat`: New feature for the user (not a part of the code, or ci, ...).
- `fix`: Bugfix for the user (not a fix to build something, ...).
- `change`: Modifying an existing visual UI instance. Such as a component or a feature.
- `refactor`: Restructuring existing code without changing its external behavior or visual UI. Typically to improve readability, maintainability, and performance.
- `test`: New tests or changes to existing tests. Does not change the production code.
- `upcoming`: A new feature that is in progress, not visible to users yet, and usually behind a feature flag.

**Example:** `feat: [M3-1234] - Allow user to view their login history`

---

nikhagra-akamai commented on packages/manager/src/features/CloudPulse/Alerts/AlertsListing/AlertListTable.tsx:
```diff
@@ -1,22 +1,24 @@
-import { Grid, TableBody, TableHead } from '@mui/material';
+import { Box } from '@linode/ui';
+import { groupByTags, sortGroups } from '@linode/utilities';
+import { Grid, TableBody, TableHead, TableRow } from '@mui/material';
```
Grid is deprecating lets start using grid 2

nikhagra-akamai commented on packages/utilities/src/types/ManagerPreferences.ts:
```diff
@@ -22,6 +22,7 @@ export type ThemeChoice = 'dark' | 'light' | 'system';
 
 export type ManagerPreferences = Partial<{
   aclpPreference: AclpConfig; // Why is this type in @linode/api-v4?
+  alerts_group_by_tag: boolean;
```
```suggestion
  alertsGroupByTag: boolean;
```

nikhagra-akamai commented on packages/manager/src/features/CloudPulse/Alerts/AlertsListing/AlertListing.tsx:
```diff
@@ -236,10 +242,12 @@ export const AlertListing = () => {
       </Box>
       <AlertsListTable
         alerts={getAlertsList}
+        alertsGroupedByTag={(preference as boolean) ?? false}
```
why type casting?

ankita-akamai replied:
This is because we are trying to generalize the usepreferencestoggle function, preference can be anything. We can also publish this somewhere in some other generalized utils file for everyone to use it. let's discuss this on next call.

venkymano-akamai replied:
@ankita-akamai , do we need a brainstorming session to avoid this, typecasting is becoming a real problem for us

nikhagra-akamai commented on packages/manager/src/features/CloudPulse/Alerts/AlertsListing/AlertListing.tsx:
```diff
@@ -236,10 +242,12 @@ export const AlertListing = () => {
         isLoading={isLoading}
         scrollToElement={() => scrollToElement(topRef.current ?? null)}
         services={getServicesList}
+        toggleAlertsGroupedByTag={toggleAlertsGroupedByTag as () => boolean}
```
same here

venkymano-akamai replied:
@ankita-akamai , why can't we avoid this...since this is a callback function and not preferences

nikhagra-akamai commented on packages/manager/src/features/CloudPulse/Alerts/AlertsListing/GroupedAlerts.tsx:
```diff
@@ -0,0 +1,125 @@
+  /**
+   *  Callback function to handle enabling or disabling an alert
+   */
+  handleEnableDisable: (alert: Alert) => void;
```
```suggestion
  handleStatusChange: (alert: Alert) => void;
```

much better name

ankita-akamai replied:
This name is not new to this pr, already approved earlier but agree that this is better. Will update this.

nikhagra-akamai commented on packages/manager/src/features/CloudPulse/Alerts/AlertsListing/UngroupedAlerts.tsx:
```diff
@@ -0,0 +1,55 @@
+  /**
+   * Callback function to handle enabling or disabling an alert
+   */
+  handleEnableDisable: (alert: Alert) => void;
```
same here

nikhagra-akamai reviewed (changes requested):
will review more in detail tomorrow

venkymano-akamai commented on packages/manager/src/factories/cloudpulse/alerts.ts:
```diff
@@ -104,7 +104,7 @@ export const alertFactory = Factory.Sync.makeFactory<Alert>({
   severity: 0,
   status: 'enabled',
-  tags: ['tag1', 'tag2'],
+  tags: [],
```
??

venkymano-akamai commented on packages/manager/src/features/CloudPulse/Alerts/AlertsListing/UngroupedAlerts.tsx:
```diff
@@ -0,0 +1,55 @@
+  services: Item<string, AlertServiceType>[];
+}
+
+export const UngroupedAlerts = ({
```
```suggestion
export const AlertsTable = ({
```

venkymano-akamai commented on packages/manager/src/features/CloudPulse/Alerts/AlertsListing/GroupedAlerts.tsx:
```diff
@@ -0,0 +1,136 @@
+  services: Item<string, AlertServiceType>[];
+}
+
+export const AlertsGroupedByTag = ({
```
```suggestion
export const GroupedAlertsTable = ({
```

venkymano-akamai commented on packages/manager/src/features/CloudPulse/Alerts/AlertsListing/GroupedAlerts.tsx:
```diff
@@ -0,0 +1,136 @@
+  );
+
+  return (
+    <React.Fragment>
```
```suggestion
    <>
```

venkymano-akamai commented on packages/manager/src/features/CloudPulse/Alerts/AlertsListing/GroupedAlerts.tsx:
```diff
@@ -0,0 +1,136 @@
+                    </StyledTagHeader>
+                  </TableCell>
+                </StyledTagHeaderRow>
+                {paginatedTagAlerts.map((alert: Alert) => (
```
```suggestion
                {paginatedTagAlerts.map((alert) => (
```

It can infer

ankita-akamai commented:
@venkymano-akamai @nikhagra-akamai I addressed all the comments( except two which i left unresolved and gave reply) and added test cases for the new components, fixed the broken ones.

venkymano-akamai commented on packages/manager/src/features/CloudPulse/Alerts/AlertsListing/AlertListTable.test.tsx:
```diff
@@ -66,9 +79,11 @@ describe('Alert List Table test', () => {
             updated,
           }),
         ]}
+        alertsGroupedByTag={false}
```
@ankita-akamai , why we need pass **false**, by default is it true?

ankita-akamai replied:
no, its false only, we can remove this

venkymano-akamai commented on packages/manager/src/features/CloudPulse/Alerts/AlertsListing/AlertListTable.tsx:
```diff
@@ -27,6 +29,10 @@ export interface AlertsListTableProps {
+  /**
+   * The current state of the alerts grouped by tag
+   */
+  alertsGroupedByTag: boolean;
```
```suggestion
  alertsGroupedByTag?: boolean;
```

why can't we make this optional property?

ankita-akamai replied:
what will we achieve by making this optional?

venkymano-akamai replied:
@ankita-akamai , we don't need pass this property unnecessarily....if it is required we need to pass either true or false

venkymano-akamai commented on packages/manager/src/features/CloudPulse/Alerts/AlertsListing/AlertListTable.tsx:
```diff
@@ -43,10 +49,22 @@ export interface AlertsListTableProps {
+  /**
+   * The callback to toggle the alerts grouped by tag
+   */
+  toggleAlertsGroupedByTag: () => boolean;
```
even this also we can make it optional

venkymano-akamai commented on packages/manager/src/features/CloudPulse/Alerts/AlertsListing/GroupedAlerts.tsx:
```diff
@@ -0,0 +1,140 @@
+  services: Item<string, AlertServiceType>[];
+}
+
+export const GroupedAlertsTable = ({
```
@ankita-akamai , file should have same name as the component

venkymano-akamai commented on packages/manager/src/features/CloudPulse/Alerts/AlertsListing/UngroupedAlerts.tsx:
```diff
@@ -0,0 +1,55 @@
+  services: Item<string, AlertServiceType>[];
+}
+
+export const AlertsTable = ({
```
@ankita-akamai , same here...

venkymano-akamai commented on packages/manager/src/features/CloudPulse/Utils/UserPreference.ts:
```diff
@@ -75,3 +75,50 @@ export const useAclpPreference = (): AclpPreferenceObject => {
+ * @param key key of the preference to be toggled
+ * @param options options for the preference
+ */
+export const usePreferencesToggle = (
```
@ankita-akamai , do we need to store this preferences in our aclp object...**what is the key here**?

venkymano-akamai commented on packages/utilities/src/types/ManagerPreferences.ts:
```diff
@@ -22,6 +22,7 @@ export type ThemeChoice = 'dark' | 'light' | 'system';
 
 export type ManagerPreferences = Partial<{
   aclpPreference: AclpConfig; // Why is this type in @linode/api-v4?
+  alertsGroupByTag: boolean;
```
```suggestion
  aclpAlertsGroupByTag: boolean;
```

Because there is already some existing alerts in linode...lets put aclp here

venkymano-akamai reviewed (changes requested):
@ankita-akamai , added some more comments, 

But this has become a big PR now, need split

ankita-akamai commented:
> @ankita-akamai , added some more comments,
> 
> But this has become a big PR now, need split

I will split the PR while raising to linode, new UTs can go in separate pr. As for aclp dev's sake, since it has already been reviewed, it can go like this.

venkymano-akamai commented on packages/manager/src/features/CloudPulse/Alerts/AlertsListing/AlertsTable.tsx:
```diff
@@ -0,0 +1,55 @@
+  services,
+}: UngroupedAlertsProps) => {
+  return (
+    <React.Fragment>
```
```suggestion
    <>
```

venkymano-akamai commented on packages/manager/src/features/CloudPulse/Alerts/AlertsListing/AlertsTable.tsx:
```diff
@@ -0,0 +1,55 @@
+          services={services}
+        />
+      ))}
+    </React.Fragment>
```
```suggestion
    </>
```

venkymano-akamai reviewed (approved):
LGTM

@ankita-akamai , lets use <></> instead of <React.Fragment> when we don't need to pass any keys or props
